genie-web/src/main/java/com/netflix/genie/web/services/impl/JobLaunchServiceImpl.java

```diff
@@ -0,0 +1,150 @@
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.services.impl;
+
+import com.google.common.collect.Sets;
+import com.netflix.genie.common.dto.JobStatus;
+import com.netflix.genie.web.agent.launchers.AgentLauncher;
+import com.netflix.genie.web.data.services.JobPersistenceService;
+import com.netflix.genie.web.dtos.JobSubmission;
+import com.netflix.genie.web.dtos.ResolvedJob;
+import com.netflix.genie.web.exceptions.checked.AgentLaunchException;
+import com.netflix.genie.web.services.JobLaunchService;
+import com.netflix.genie.web.services.JobResolverService;
+import com.netflix.genie.web.util.MetricsUtils;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.Nonnull;
+import javax.validation.Valid;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Default implementation of the {@link JobLaunchService}.
+ *
+ * @author tgianos
+ * @since 4.0.0
+ */
+@Slf4j
+public class JobLaunchServiceImpl implements JobLaunchService {
+
+    private static final String LAUNCH_JOB_TIMER = "genie.services.jobLaunch.launchJob.timer";
+
+    private final JobPersistenceService jobPersistenceService;
+    private final JobResolverService jobResolverService;
+    private final AgentLauncher agentLauncher;
+    private final MeterRegistry registry;
+
+    /**
+     * Constructor.
+     *
+     * @param jobPersistenceService {@link JobPersistenceService} implementation to save job data
+     * @param jobResolverService    {@link JobResolverService} implementation used to resolve job details
+     * @param agentLauncher         {@link AgentLauncher} implementation to launch agents
+     * @param registry              {@link MeterRegistry} metrics repository
+     */
+    public JobLaunchServiceImpl(
+        final JobPersistenceService jobPersistenceService,
+        final JobResolverService jobResolverService,
+        final AgentLauncher agentLauncher,
+        final MeterRegistry registry
+    ) {
+        this.jobPersistenceService = jobPersistenceService;
+        this.jobResolverService = jobResolverService;
+        this.agentLauncher = agentLauncher;
+        this.registry = registry;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @Nonnull
+    public String launchJob(@Valid final JobSubmission jobSubmission) throws AgentLaunchException {
+        final long start = System.nanoTime();
+        final Set<Tag> tags = Sets.newHashSet();
+        try {
+            /*
+             * Steps:
+             *
+             * 1. Save the job information
+             * 2. Attempt to resolve the job information (includes saving)
```

Comment:
I'm always leaning towards separating, in this case resolve and save could be split.
I don't know the specifics so you may have a good reason, but in general it seems to make it easier to maintain, test, etc.

Reply:
Yeah I went back and forth on this a bit. If you look at the `JobCoordinatorServiceImpl` from V3 it's a pretty big mess due to not encapsulating these things behind a chain of composite API calls. We'd have to get the data back in a DTO form and then go and fetch a bunch of data again from the database and then save. There effectively were a whole bunch of transactions:
- Initial Job save
- search for job spec information
- Retrieval of cluster
- Retrieval of command
- Retrieval of applications
- Saving runtime information
- Updating status
- and probably others I've forgotten without looking at the code

This way the `JobResolverService` itself is a higher level API that abstracts some of those details and allows us to limit it to a single transaction scope and save database calls, increase speed, etc. It still delegates to an internal `JobPersistenceService` implementation so the separation is there but the `resolveJob` API does do a couple of things. It is documented at the API level on `resolveJob` and there is another `resolveJob` API that doesn't do the save if that is desired. If we really want to be exacting about it we could create another middle service which wraps the resolver service and the persistence in order to keep them really separate. I felt this was over designing and went with the best performance option over purity of code in this instance.